src/install/dracut-install.c

```diff
@@ -1515,7 +1515,105 @@ static bool check_module_path(const char *path)
+
+                                        _cleanup_kmod_module_unref_list_ struct kmod_list *list = NULL;
+                                        struct kmod_list *l = NULL;
+
+                                        if (find_module_from_sysfs_node(ctx, supplier_path, &list) >= 0) {
+                                                kmod_list_foreach(l, list) {
+                                                        struct kmod_module *mod = kmod_module_get_module(l);
+                                                        char *supplier_module = strdup(kmod_module_get_name(mod));
+                                                        kmod_module_unref(mod);
+                                                        hashmap_put(suppliers, supplier_module, supplier_module);
+                                                }
+                                        }
+                                }
+                        }
+                }
+                closedir(d);
+        }
+}
+
+static void find_suppliers_for_module(struct kmod_ctx *ctx, Hashmap *suppliers, const char *module)
+{
+        _cleanup_fts_close_ FTS *fts;
+        char *paths[] = { "/sys/devices/platform", NULL };
+        fts = fts_open(paths, FTS_NOSTAT | FTS_PHYSICAL, NULL);
+
+        for (FTSENT *ftsent = fts_read(fts); ftsent != NULL; ftsent = fts_read(fts)) {
+                if (strcmp(ftsent->fts_name, "modalias") == 0) {
+                        char sysfs_node[PATH_MAX] = {0};
+                        memcpy(sysfs_node, ftsent->fts_parent->fts_path, ftsent->fts_parent->fts_pathlen);
+                        sysfs_node[ftsent->fts_parent->fts_pathlen] = '\0';
+
+                        _cleanup_kmod_module_unref_list_ struct kmod_list *list = NULL;
+                        struct kmod_list *l = NULL;
+
+                        if (find_module_from_sysfs_node(ctx, sysfs_node, &list) < 0)
+                                continue;
+
+                        kmod_list_foreach(l, list) {
+                                _cleanup_kmod_module_unref_ struct kmod_module *mod = kmod_module_get_module(l);
+                                const char *name = kmod_module_get_name(mod);
+
+                                if (strcmp(name, module) == 0) {
+                                        find_suppliers_for_sys_node(ctx, suppliers, sysfs_node);
+
+                                        // Also add suppliers of parent nodes
+                                        FTSENT *parent = ftsent->fts_parent;
+                                        while(parent->fts_level >= 1) {
+                                                memcpy(sysfs_node, parent->fts_path, parent->fts_pathlen);
+                                                sysfs_node[parent->fts_pathlen] = '\0';
+
+                                                find_suppliers_for_sys_node(ctx, suppliers, sysfs_node);
```

Comment:
AFAICT this processes `ftsent->fts_parent` twice (once raw, aliased as sysfs_node, the other time in the first loop iteration)?